File: codegen/ui_code_generator.py
"""ASTRA UI code generator - generates pytest Playwright UI test files."""
from __future__ import annotations

from pathlib import Path
from typing import Any, Dict, List, Optional

from schemas.field_schema import FieldSchema, ResolvedField
from core.search.a_star_engine import AStarResult
from utils.logger import logger

_TESTS_DIR = Path(__file__).parent.parent / "tests" / "ui"


class UiCodeGenerator:
    def __init__(self, schema: FieldSchema, astar_result: AStarResult) -> None:
        self.schema = schema
        self.astar_result = astar_result

    def generate(self) -> List[str]:
        """Generate 3 UI test files: positive, negative, full."""
        files_generated: List[str] = []

        test_files = [
            ("test_positive.py", self._generate_positive_test),
            ("test_negative.py", self._generate_negative_test),
            ("test_full.py", self._generate_full_test),
        ]

        _TESTS_DIR.mkdir(parents=True, exist_ok=True)

        for filename, generator in test_files:
            content = generator()
            path = _TESTS_DIR / filename
            path.write_text(content, encoding="utf-8")
            files_generated.append(str(path))
            logger.codegen(f"Generated {filename}")

        return files_generated

    def _generate_positive_test(self) -> str:
        steps = self._build_steps()
        return f'''"""UI test: Positive registration flow"""
import pytest
from playwright.sync_api import Page


@pytest.mark.ui
@pytest.mark.positive
def test_registration_positive(authenticated_page: Page):
    """Test successful user registration with valid data."""
    page = authenticated_page
    page.goto("{self.schema.target_endpoint}")

    # Fill form with valid data
{steps}

    # Submit
    page.click('[type=submit], button:has-text("Submit"), button:has-text("Register")')
    page.wait_for_load_state("networkidle")

    # Assert success
    page.wait_for_url("**/success**")
    assert page.is_visible(".success-message, #success")
'''

    def _generate_negative_test(self) -> str:
        return '''"""UI test: Negative registration flows"""
import pytest
from playwright.sync_api import Page


@pytest.mark.ui
@pytest.mark.negative
def test_registration_missing_required(authenticated_page: Page):
    """Test registration fails with missing required field."""
    page = authenticated_page
    page.goto("{self.schema.target_endpoint}")

    # Fill only optional fields
    page.fill('[placeholder*="First"], #firstName', "John")
    page.click('[type=submit], button:has-text("Submit")')

    # Assert error
    assert page.is_visible(".error, [data-testid=error]")
    assert not page.is_visible(".success-message")


@pytest.mark.ui
@pytest.mark.negative
def test_registration_invalid_email(authenticated_page: Page):
    """Test registration fails with invalid email."""
    page = authenticated_page
    page.goto("{self.schema.target_endpoint}")

    page.fill('[type=email], [name=email]', "not-an-email")
    page.click('[type=submit], button:has-text("Submit")')

    assert page.is_visible(".error, [data-testid=error]")
'''

    def _generate_full_test(self) -> str:
        return '''"""UI test: Full comprehensive registration"""
import pytest
from playwright.sync_api import Page


@pytest.mark.ui
@pytest.mark.e2e
def test_registration_full_flow(authenticated_page: Page):
    """Test complete registration with all fields and validations."""
    page = authenticated_page
    page.goto("{self.schema.target_endpoint}")

    # Verify all required fields are visible
    assert page.is_visible('[name=firstName]')
    assert page.is_visible('[name=lastName]')
    assert page.is_visible('[type=email]')

    # Fill all fields
    page.fill('[name=firstName]', "John")
    page.fill('[name=lastName]', "Doe")
    page.fill('[type=email]', "john.doe@example.com")
    page.fill('[type=password]', "Secure@123")

    # Handle conditional fields if present
    if page.query_selector('[name=country]'):
        page.select_option('[name=country]', "US")

    # Accept terms
    if page.query_selector('[name=terms]'):
        page.check('[name=terms]')

    # Submit
    page.click('[type=submit], button:has-text("Register")')
    page.wait_for_load_state("networkidle")

    # Assert success
    assert page.url.endswith("/success") or "success" in page.content()
'''

    def _build_steps(self) -> str:
        """Generate type-aware Playwright fill/select/check actions."""
        steps = []
        for field in self.astar_result.path:
            selector = self._build_fallback_selector(field)
            value = self._generate_test_value(field)
            if field.type == "checkbox":
                steps.append(f'    page.check("{selector}")')
            elif field.type == "select":
                steps.append(f'    page.select_option("{selector}", "{value}")')
            elif field.type == "radio":
                steps.append(f'    page.check("{selector}")')
            else:
                steps.append(f'    page.fill("{selector}", "{value}")')
        return "\n".join(steps)

    def _build_fallback_selector(self, field: ResolvedField) -> str:
        """5-strategy selector fallback."""
        if field.selector:
            return field.selector
        strategies = [
            f'[name="{field.name}"]',
            f'#{field.name}',
            f'[data-testid="{field.name}"]',
            f'[placeholder*="{field.name}"]',
            f'label:has-text("{field.name}") + input',
        ]
        return " ,".join(strategies)

    def _generate_test_value(self, field: ResolvedField) -> str:
        """Generate type-aware test data."""
        mapping = {
            "email": "test@example.com",
            "password": "Secure@123",
            "tel": "+1234567890",
            "number": "42",
            "date": "2000-01-01",
        }
        return mapping.get(field.type, f"test_{field.name}")

File: codegen/uiCodeGenerator.ts
// ═══════════════════════════════════════════════════════════════
// codegen/uiCodeGenerator.ts
// ASTRA Framework — UI Playwright Test Code Generator
// Consumes A* search result → generates ready-to-run .ts test files
// ═══════════════════════════════════════════════════════════════
//Key highlights — generates real runnable Playwright .ts files:
//
// 3 test files generated per run — positive.spec.ts (valid data), negative.spec.ts (invalid data), full.spec.ts (all fields)
// buildFieldAction() — type-aware Playwright actions:
//
// dropdown → selectOption()
// checkbox → check() / uncheck() with state check
// radio → check() on value selector
// file → setInputFiles()
// everything else → clear() + fill()
//
//
// buildFallbackSelector() — 5-strategy selector: [name] → [id] → [data-testid] → [placeholder*=] → [aria-label*=]
// Multi-step form support — detects step transitions, inserts waitForSelector('[data-step="N"]') automatically
// buildGoalAssertion() — negative mode asserts .error/[role="alert"] visible, positive mode asserts success URL/text/selector
// Every step includes f-score comment — full A* transparency in generated code
// Auto-header comment: schema ID, mode, timestamp, path length, iterations — fully traceable

import * as fs   from "fs-extra";
import * as path from "path";
import { ENV }   from "../utils/envLoader";
import { logger } from "../utils/logger";
import { AStarEngine }  from "../core/search/aStarEngine";
import { AStarResult, AStarNode } from "../core/search/aStarEngine";
import { FieldSchema }  from "../schemas/fieldSchema.interface";

// ═══════════════════════════════════════════════════════════════
// Interfaces
// ═══════════════════════════════════════════════════════════════
export interface UiCodeGenResult {
  positiveTestFile: string;
  negativeTestFile: string;
  fullTestFile:     string;
  savedTo:          string[];
}

// ═══════════════════════════════════════════════════════════════
// UiCodeGenerator
// ═══════════════════════════════════════════════════════════════
export class UiCodeGenerator {

  private readonly schema:  FieldSchema;
  private readonly engine:  AStarEngine;
  private readonly outDir:  string;

  constructor(schema: FieldSchema) {
    this.schema = schema;
    this.engine = new AStarEngine(schema);
    this.outDir = path.resolve(__dirname, "../tests/generated/ui");
  }

  // ═══════════════════════════════════════════════════════════
  // generate — runs A* and produces all test files
  // ═══════════════════════════════════════════════════════════
  async generate(): Promise<UiCodeGenResult> {
    logger.divider("UI Code Generator — Starting");
    await fs.ensureDir(this.outDir);

    // ─── Run A* searches ──────────────────────────────────────
    logger.codegen("Running A* positive search...");
    const positiveResult = this.engine.search({ includeOptional: false });
    logger.codegen(this.engine.getSearchSummary(positiveResult));

    logger.codegen("Running A* negative search...");
    const negativeResult = this.engine.searchNegative();

    logger.codegen("Running A* full search (with optional)...");
    const fullResult = this.engine.searchWithOptional();

    // ─── Generate test files ──────────────────────────────────
    const positiveCode = this.generateTestFile(positiveResult, "positive");
    const negativeCode = this.generateTestFile(negativeResult, "negative");
    const fullCode     = this.generateTestFile(fullResult,     "full");

    // ─── Save files ───────────────────────────────────────────
    const schemaId    = this.schema.schemaId.replace(/[^a-z0-9]/gi, "_");
    const positivePath = path.join(this.outDir, `${schemaId}.positive.spec.ts`);
    const negativePath = path.join(this.outDir, `${schemaId}.negative.spec.ts`);
    const fullPath     = path.join(this.outDir, `${schemaId}.full.spec.ts`);

    await fs.writeFile(positivePath, positiveCode, "utf-8");
    await fs.writeFile(negativePath, negativeCode, "utf-8");
    await fs.writeFile(fullPath,     fullCode,     "utf-8");

    logger.codegen(`✅ Positive test  → ${positivePath}`);
    logger.codegen(`✅ Negative test  → ${negativePath}`);
    logger.codegen(`✅ Full test      → ${fullPath}`);

    return {
      positiveTestFile: positivePath,
      negativeTestFile: negativePath,
      fullTestFile:     fullPath,
      savedTo:          [positivePath, negativePath, fullPath],
    };
  }

  // ═══════════════════════════════════════════════════════════
  // generateTestFile — builds complete Playwright spec file
  // ═══════════════════════════════════════════════════════════
  private generateTestFile(
    result:  AStarResult,
    mode:    "positive" | "negative" | "full"
  ): string {
    const schemaId    = this.schema.schemaId;
    const goalUi      = this.schema.goalCondition.ui;
    const isNegative  = mode === "negative";
    const testTitle   = this.buildTestTitle(mode);
    const steps       = this.buildSteps(result.path, isNegative);
    const goalAssert  = this.buildGoalAssertion(goalUi, isNegative);
    const timestamp   = new Date().toISOString();

    return `// ═══════════════════════════════════════════════════════
// AUTO-GENERATED BY ASTRA FRAMEWORK
// Schema  : ${schemaId}
// Mode    : ${mode.toUpperCase()}
// Generated: ${timestamp}
// A* Path : ${result.path.length} steps | Iterations: ${result.iterations}
// DO NOT EDIT MANUALLY — re-run codegen to regenerate
// ═══════════════════════════════════════════════════════

import { test, expect, Page } from "@playwright/test";

// ─── Test Configuration ────────────────────────────────
const BASE_URL    = process.env.BASE_URL    || "${ENV.BASE_URL}";
const LOGIN_URL   = process.env.LOGIN_URL   || "${ENV.LOGIN_URL}";
const TARGET_URL  = process.env.TARGET_PAGE_URL || "${ENV.TARGET_PAGE_URL}";
const USERNAME    = process.env.APP_USERNAME || "${ENV.APP_USERNAME}";
const PASSWORD    = process.env.APP_PASSWORD || "***";

// ─── Test Suite ────────────────────────────────────────
test.describe("${schemaId} — ${mode.toUpperCase()} Tests", () => {

  test.beforeEach(async ({ page }) => {
    await loginToApp(page);
    await page.goto(TARGET_URL, { waitUntil: "networkidle" });
    await page.waitForLoadState("domcontentloaded");
  });

  // ─── ${testTitle} ───────────────────────────────────
  test("${testTitle}", async ({ page }) => {

${steps}

${goalAssert}

  });

});

// ─── Login Helper ──────────────────────────────────────
async function loginToApp(page: Page): Promise<void> {
  await page.goto(LOGIN_URL, { waitUntil: "networkidle" });

  const usernameSelectors = [
    'input[name="username"]', 'input[name="email"]',
    'input[type="email"]',   'input[id*="user"]',
  ];
  const passwordSelectors = [
    'input[name="password"]', 'input[type="password"]',
  ];
  const submitSelectors = [
    'button[type="submit"]', 'button:has-text("Login")',
    'button:has-text("Sign in")',
  ];

  for (const sel of usernameSelectors) {
    if (await page.locator(sel).isVisible().catch(() => false)) {
      await page.locator(sel).fill(USERNAME);
      break;
    }
  }
  for (const sel of passwordSelectors) {
    if (await page.locator(sel).isVisible().catch(() => false)) {
      await page.locator(sel).fill(PASSWORD);
      break;
    }
  }
  for (const sel of submitSelectors) {
    if (await page.locator(sel).isVisible().catch(() => false)) {
      await page.locator(sel).click();
      break;
    }
  }

  await page.waitForLoadState("networkidle");
}
`;
  }

  // ═══════════════════════════════════════════════════════════
  // buildSteps — generates Playwright fill/select/click actions
  // ═══════════════════════════════════════════════════════════
  private buildSteps(nodes: AStarNode[], isNegative: boolean): string {
    const lines: string[] = [];
    let   lastStep = -1;

    lines.push(`    // ─── A* Generated Steps [${isNegative ? "NEGATIVE" : "POSITIVE"} mode] ─────`);

    for (const node of nodes) {
      const { field, value, fieldPath } = node;

      // ─── Step transition comment ────────────────────────
      const stepIndex = field.stepIndex ?? 0;
      if (stepIndex !== lastStep && stepIndex > 0) {
        lines.push(`\n    // ── Step ${stepIndex + 1} ──────────────────────────`);
        lines.push(`    await page.waitForSelector('[data-step="${stepIndex}"], .step-${stepIndex}').catch(() => {});`);
        lastStep = stepIndex;
      }

      // ─── Section comment ────────────────────────────────
      lines.push(`\n    // [${field.sectionName}] ${fieldPath} | mandatory=${field.mandatory} | f=${node.fScore.toFixed(1)}`);

      // ─── Generate action based on field type ────────────
      const action = this.buildFieldAction(field.type, field.selector, fieldPath, value);
      lines.push(...action);
    }

    // ─── Submit action ───────────────────────────────────
    lines.push(`\n    // ── Submit Form ──────────────────────────────`);
    lines.push(`    const submitBtn = page.locator('button[type="submit"], input[type="submit"], button:has-text("Submit"), button:has-text("Save"), button:has-text("Create")').first();`);
    lines.push(`    await expect(submitBtn).toBeVisible({ timeout: 5000 });`);
    lines.push(`    await submitBtn.click();`);

    return lines.join("\n");
  }

  // ═══════════════════════════════════════════════════════════
  // buildFieldAction — generates specific Playwright action per type
  // ═══════════════════════════════════════════════════════════
  private buildFieldAction(
    type:      string,
    selector:  string | undefined,
    fieldPath: string,
    value:     string
  ): string[] {
    // ─── Build best available selector ──────────────────
    const fieldName  = fieldPath.split(".").pop() ?? fieldPath;
    const sel        = selector
      ?? this.buildFallbackSelector(fieldName, type);

    const locator = `page.locator(${JSON.stringify(sel)}).first()`;
    const lines: string[] = [];

    switch (type) {
      case "dropdown":
        lines.push(`    await expect(${locator}).toBeVisible({ timeout: 5000 });`);
        lines.push(`    await ${locator}.selectOption(${JSON.stringify(value)});`);
        break;

      case "checkbox":
        lines.push(`    await expect(${locator}).toBeVisible({ timeout: 5000 });`);
        if (value === "true") {
          lines.push(`    if (!await ${locator}.isChecked()) await ${locator}.check();`);
        } else {
          lines.push(`    if (await ${locator}.isChecked()) await ${locator}.uncheck();`);
        }
        break;

      case "radio":
        lines.push(`    await page.locator(\`input[type="radio"][value="${value}"]\`).check();`);
        break;

      case "date":
        lines.push(`    await expect(${locator}).toBeVisible({ timeout: 5000 });`);
        lines.push(`    await ${locator}.fill(${JSON.stringify(value)});`);
        break;

      case "file":
        lines.push(`    await ${locator}.setInputFiles(${JSON.stringify(value)});`);
        break;

      case "textarea":
        lines.push(`    await expect(${locator}).toBeVisible({ timeout: 5000 });`);
        lines.push(`    await ${locator}.fill(${JSON.stringify(value)});`);
        break;

      default:
        // text | email | password | phone | number | url
        lines.push(`    await expect(${locator}).toBeVisible({ timeout: 5000 });`);
        lines.push(`    await ${locator}.clear();`);
        lines.push(`    await ${locator}.fill(${JSON.stringify(value)});`);
        break;
    }

    return lines;
  }

  // ═══════════════════════════════════════════════════════════
  // buildFallbackSelector — generates best-guess CSS selector
  // ═══════════════════════════════════════════════════════════
  private buildFallbackSelector(fieldName: string, type: string): string {
    const selectors = [
      `[name="${fieldName}"]`,
      `[id="${fieldName}"]`,
      `[data-testid="${fieldName}"]`,
      `[placeholder*="${fieldName}" i]`,
      `[aria-label*="${fieldName}" i]`,
    ];

    if (type === "dropdown") return `select[name="${fieldName}"], [name="${fieldName}"]`;
    if (type === "checkbox") return `input[type="checkbox"][name="${fieldName}"]`;
    if (type === "radio")    return `input[type="radio"][name="${fieldName}"]`;

    return selectors[0];
  }

  // ═══════════════════════════════════════════════════════════
  // buildGoalAssertion — generates success assertion
  // ═══════════════════════════════════════════════════════════
  private buildGoalAssertion(
    goalUi:     FieldSchema["goalCondition"]["ui"],
    isNegative: boolean
  ): string {
    const lines: string[] = [];
    lines.push(`\n    // ── Goal Assertion ──────────────────────────`);

    if (isNegative) {
      lines.push(`    // Negative test — expect validation errors, NOT success`);
      lines.push(`    await expect(page.locator('.error, .alert-error, [role="alert"], .invalid-feedback').first())`);
      lines.push(`      .toBeVisible({ timeout: 5000 });`);
      return lines.join("\n");
    }

    if (goalUi?.urlContains) {
      lines.push(`    await expect(page).toHaveURL(/${goalUi.urlContains}/, { timeout: ${ENV.ASTAR_GOAL_TIMEOUT} });`);
    }

    if (goalUi?.successSelector) {
      lines.push(`    await expect(page.locator(${JSON.stringify(goalUi.successSelector)}).first())`);
      lines.push(`      .toBeVisible({ timeout: ${ENV.ASTAR_GOAL_TIMEOUT} });`);
    }

    if (goalUi?.successText) {
      lines.push(`    await expect(page.locator(\`text=${goalUi.successText}\`).first())`);
      lines.push(`      .toBeVisible({ timeout: ${ENV.ASTAR_GOAL_TIMEOUT} });`);
    }

    if (!goalUi?.urlContains && !goalUi?.successSelector && !goalUi?.successText) {
      lines.push(`    // Fallback: no error visible = success`);
      lines.push(`    await expect(page.locator('.error, .alert-error, [role="alert"]').first())`);
      lines.push(`      .not.toBeVisible({ timeout: 5000 }).catch(() => {});`);
    }

    return lines.join("\n");
  }

  // ═══════════════════════════════════════════════════════════
  // buildTestTitle
  // ═══════════════════════════════════════════════════════════
  private buildTestTitle(mode: string): string {
    const titles: Record<string, string> = {
      positive: "should submit form successfully with valid mandatory fields (A* happy path)",
      negative: "should show validation errors with invalid field values (A* negative path)",
      full:     "should submit form with all fields filled — mandatory + optional (A* full path)",
    };
    return titles[mode] ?? `${mode} test`;
  }
}

// ═══════════════════════════════════════════════════════════════
// Direct execution: npm run codegen:ui
// ═══════════════════════════════════════════════════════════════
async function main(): Promise<void> {
  const schemaPath = path.resolve(__dirname, "../schemas/ui/autoGeneratedSchema.json");

  if (!await fs.pathExists(schemaPath)) {
    logger.error(`UI schema not found at ${schemaPath}`);
    logger.error("Run preflight first: npm run preflight");
    process.exit(1);
  }

  const schema: FieldSchema = await fs.readJson(schemaPath);
  const generator = new UiCodeGenerator(schema);
  const result    = await generator.generate();

  logger.info("\n✅ UI Code Generation complete:");
  result.savedTo.forEach((f) => logger.info(`   → ${f}`));
}

if (require.main === module) {
  main().catch((err) => {
    logger.error(`UI codegen failed: ${err}`);
    process.exit(1);
  });
}